src/server/requesters/AzureDevOpsRequester.ts

```diff
@@ -0,0 +1,269 @@
+
+        if (response.status !== 200) {
+            return {
+                ok: false,
+                error: "Release with id " + request.requestMetadata.releaseId + " does not exist!"
+            }
+        }
+
+        if (response.data.operationStatus != 'phaseInProgress') {
+            return {
+                ok: false,
+                error: "Release with id " + request.requestMetadata.releaseId + " is not in progress!"
+            }
+        }
+
+        return {
+            ok: true
+        }
+    }
+    isOTPRequestValidForRequester(request: OTPRequest<unknown, unknown>): Promise<OTPRequest<AzDOOTPRequestMetadata, unknown> | null> {
+        const result = validateMetadataObject(request);
+        return result.error ? null : request as any;
+    }
+
+    async getRequestInformationToPassOn(request: OTPRequest<AzDOOTPRequestMetadata, unknown>): Promise<RequestInformation> {
+        const { project } = request;
+
+        const axios = getAxiosForConfig(project.requester_AzureDevOps!);
+        const release = await axios.get<AzDORelease>(`release/releases/${request.requestMetadata.releaseId}`);
+
+        return {
+            description: `Azure DevOps Release ${project.repoOwner}/${project.repoName}#${request.requestMetadata.releaseId}`,
+            url: release.data._links.web.href
+        }
+    }
+
+    validateProofForRequest(
+        request: OTPRequest<AzDOOTPRequestMetadata, unknown>,
+        config: AzureDevOpsRequesterConfig
+    ): Promise<boolean> {
+
+        async function attemptValidateProof(attempts: number) {
+            if (attempts <= 0) {
+                return false;
+            }
+
+            const again = async () => {
+                await new Promise(r => setTimeout(r, 5000));
+                return attemptValidateProof(attempts - 1);
+            };
+
+            return getLogs(config, request.requestMetadata.releaseId)
+                .then(({ logs }) => logs.includes(request.proof) ? true : again())
+                .catch(e => {
+                    return again();
+                });
+        }
+
+        return attemptValidateProof(3);
+    }
+
+}
+
+/**
+ * Get the logs from an Azure Dev Ops Release pipeline. They come in a zip file with multiple log files in it - 1 per phase task. 
+ * Reject if could not reach the organization/project/releaseDef or there were no runs/logs in that release 
+ * @param config  
+ * @param releaseId 
+ * @example 
+ * // See release definition https://dev.azure.com/gparlakov/Scuri/_release?definitionId=1&view=mine&_a=releases
+ * // and release run https://dev.azure.com/gparlakov/Scuri/_releaseProgress?_a=release-environment-logs&releaseId=58&environmentId=87 
+ * config.organizationName === 'gparlakov';
+ * config.projectName === 'scuri';
+ * config.accessToken = 'easasdasdasdasdasda12d2312d13ed1d'; 
+ * releaseId === 58  
+ */
+export function getLogs(
```

Comment:
Can we do this entirely in memory instead of writing to disk, I've seen zip-based file system attacks before and I don't want to be susceptible to an out-of-zip write attack or something.

There are a few nodejs zip libraries you could probably use to do the work in-memory

Reply:
Potentially just using the `fromBuffer` method of this zip library you're currently using.

https://www.npmjs.com/package/yauzl#frombufferbuffer-options-callback

Reply:
Sure, especially if there is a potential security vulnerability. 
I wanted to keep memory consumption low by not keeping the buffer/zip stream and string at the same time.

Reply:
I think paying the in-memory price here is ok, these log zips shouldnt be too big and theyll only be in memory for a second or two.